### PR TITLE
Review update

### PR DESCRIPTION
**Pull request by priyanka-TL**

### CODE REVIEW

File: src/.env.sample [hunk 1]
```diff
@@ -120,28 +120,39 @@ ALLOWED_HOST=http://exampleDomain.com
...
 MAX_PROJECT_TASK_COUNT= 10
+
+#consumption portal name ex: self/diksha/elevate-project
+CONSUMPTION_SERVICE=self
+
+#Publish method to consumption side kafka/api
+PUBLISH_METHOD=kafka
+
+#kafka topic for publish
+PROJECT_PUBLISH_KAFKA_TOPIC=publishtopic
```
Comment (adithyadinesh0412) [inline src/.env.sample:157]: can we make it resource publish kafka topic?  @priyanka-TL

File: src/database/migrations/20240513091857-create-reviews-table.js [hunk 2]
```diff
@@ -27,7 +27,8 @@ module.exports = {
 					'APPROVED',
 					'REJECTED',
 					'PUBLISHED',
-					'REJECTED_AND_REPORTED'
+					'REJECTED_AND_REPORTED',
+					'CHANGES_UPDATED'
```
Comment (adithyadinesh0412) [inline src/database/migrations/20240513091857-create-reviews-table.js]: this migration is already executed right? @priyanka-TL

File: src/database/migrations/20240513094933-create-comments-table.js
Comment (adithyadinesh0412) [inline src/database/migrations/20240513094933-create-comments-table.js:1]: this migration is also already executed right? @priyanka-TL

File: src/services/comment.js
Comment (adithyadinesh0412) [inline src/services/comment.js:1]: why is this file removed? @priyanka-TL

File: src/services/projects.js [hunk 3]
```diff
@@ -38,7 +38,7 @@ module.exports = class ProjectsHelper {
 			let projectData = {
 				title: bodyData.title,
 				type: common.PROJECT,
-				status: common.STATUS_DRAFT,
+				status: common.RESOURCE_STATUS_DRAFT,
```
Comment (adithyadinesh0412) [inline src/services/projects.js:50]: Can we change the order of arguments in create(orgId, loggedInUserId, bodyData) to create(loggedInUserId , orgId,  bodyData) ?

File: src/services/projects.js [hunk 4]
```diff
@@ -254,13 +253,13 @@ module.exports = class ProjectsHelper {
...
 			}
 
-			if (project.status == common.STATUS_DRAFT && project.user_id !== loggedInUserId) {
-				return responses.failureResponse({
-					message: 'PROJECT_NOT_VISIBLE',
-					statusCode: httpStatusCode.bad_request,
-					responseCode: 'CLIENT_ERROR',
-				})
-			}
+			// if (project.status == common.RESOURCE_STATUS_DRAFT && project.user_id !== loggedInUserId) {
```
Comment (adithyadinesh0412) [inline src/services/projects.js]: why is this commented? @priyanka-TL

File: src/.env.sample (hunk 1 above)
Comment (priyanka-TL) [reply src/.env.sample:157]: for different resources publish topic will be different, like elevate-project can create the project, but if we want to create observation then we need to call elevate-samiksha kafka topic

File: src/services/reviews.js [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+			let isPublishResource = false
+			let updateNextLevel = false
+
+			// Handle review creation
+			if (!review) {
+				//check review type is sequential and nobody started review
+				if (reviewType === common.REVIEW_TYPE_SEQUENTIAL) {
+					const existingReview = await reviewsQueries.findOne({
+						resource_id: resourceId,
+						status: _notAllowedReviewStatus,
```
Comment (adithyadinesh0412) [inline src/services/reviews.js]: in or not-in operation not required? @priyanka-TL

File: src/services/reviews.js [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+						message: 'FAILED_TO_START_REVIEW',
+						statusCode: httpStatusCode.internal_server_error,
+						responseCode: 'CLIENT_ERROR',
+					})
+				}
+
+				await reviewResourceQueries.create(_.omit(reviewData, [common.STATUS]))
+
+				// Update resource table data
+				let filter = {
```
Comment (adithyadinesh0412) [inline src/services/reviews.js]: this is not a filter right? @priyanka-TL

File: src/services/reviews.js [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+async function handleComments(comments, resourceId, loggedInUserId) {
+	try {
+		// Normalize comments to an array if it's a single object
+		if (!Array.isArray(comments)) {
+			comments = [comments]
+		}
+		// handle adding or updating comments
+		for (const comment of comments) {
+			if (comment.id) {
+				await commentQueries.updateOne({ id: comment.id, resource_id: resourceId }, _.omit(comment, ['id']))
```
Comment (adithyadinesh0412) [inline src/services/reviews.js]: We can update only draft comments right? Should we check the status also here? @priyanka-TL

File: src/services/reviews.js [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+				{ status: common.RESOURCE_STATUS_APPROVED }
+			)
+
+			// Add or update comments
+			if (comments) {
+				await handleComments(comments, resourceId, loggedInUserId)
+			}
+
+			//check the no of approvals meets
+			const reviewsApproved = await reviewsQueries.countDistinct({
```
Comment (adithyadinesh0412) [inline src/services/reviews.js]: There is change in countDistinct funtion. Can we discuss to handle this better? @priyanka-TL

File: src/services/reviews.js [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+				resource_id: resourceId,
+				status: common.REVIEW_STATUS_APPROVED,
+			})
+
+			if (minApproval <= reviewsApproved + 1) {
```
Comment (adithyadinesh0412) [inline src/services/reviews.js]: if minApproval of a resource is 2 and a reviewer approves this resource. then
reviewsApproved = 1
reviewsApproved + 1 = 2

then minApproval <= reviewsApproved + 1 will be true and it will be published right? 
@priyanka-TL

File: src/.env.sample (hunk 1 above)
Comment (adithyadinesh0412) [reply src/.env.sample:157]: ohk got it @priyanka-TL

File: src/database/migrations/20240513094933-create-comments-table.js [hunk 10]
```diff
@@ -30,8 +30,8 @@ module.exports = {
 			},
 			status: {
 				allowNull: false,
-				type: Sequelize.ENUM('OPEN', 'RESOLVED'),
-				defaultValue: 'OPEN',
+				type: Sequelize.ENUM('OPEN', 'RESOLVED', 'DRAFT'),
```
Comment (adithyadinesh0412) [inline src/database/migrations/20240513094933-create-comments-table.js:33]: We have to update the enum manually for this right? @priyanka-TL

File: src/configs/kafka.js [hunk 11]
```diff
@@ -47,6 +47,7 @@ module.exports = async () => {
 			},
 		})
 	}
+
```
Comment (adithyadinesh0412) [inline src/configs/kafka.js]: This file is not updated in this PR right? can we remove this additional line? @priyanka-TL

**Review by adithyadinesh0412 — COMMENTED** [review]
Reviewed on 24 July : 11.48AM

File: src/controllers/v1/comments.js [hunk 12]
```diff
@@ -15,7 +15,13 @@ module.exports = class configs {
 	 */
 	async list(req) {
 		try {
-			const comments = await commentService.list(req.query.resource_id, req.decodedToken.organization_id)
+			const comments = await commentService.list(
+				req.query.resource_id,
+				req.query.page_value ? req.query.page_value : req.query.page_value,
+				req.query.context ? req.query.context : req.query.context,
+				req.decodedToken.id,
+				req.decodedToken.organization_id
```
Comment (adithyadinesh0412) [inline src/controllers/v1/comments.js:26]: service file for this controlled is deleted and im not able to see a service for this. Can u check? @priyanka-TL

File: src/database/migrations/20240513091857-create-reviews-table.js (hunk 2 above)
Comment (priyanka-TL) [reply src/database/migrations/20240513091857-create-reviews-table.js]: Updated in old migration we can add this directly

File: src/services/comment.js
Comment (priyanka-TL) [reply src/services/comment.js:1]: Added

File: src/services/reviews.js (hunk 7 above)
Comment (priyanka-TL) [reply src/services/reviews.js]: no, they can resolve it also that is not a separate API all

File: src/database/migrations/20240513094933-create-comments-table.js (hunk 10 above)
Comment (priyanka-TL) [reply src/database/migrations/20240513094933-create-comments-table.js:33]: yes

File: src/controllers/v1/comments.js (hunk 12 above)
Comment (priyanka-TL) [reply src/controllers/v1/comments.js:26]: added

File: src/.env.sample [hunk 13]
```diff
@@ -147,5 +147,14 @@ RESOURCE_TYPES="projects,observations,observation_with_rubrics,surveys,programs"
 #maximum number of tasks allowed in a project
 MAX_PROJECT_TASK_COUNT= 10
 
+#consumption portal name ex: self/diksha/elevate-project
+CONSUMPTION_SERVICE=self
+
+#Publish method to consumption side kafka/api
+PUBLISH_METHOD=kafka
```
Comment (aks30) [inline src/.env.sample]: @priyanka-TL we don't need this key, basically if API URL is given then push to API, if kafka topic is given we should push to Kafka, if both then do both.

File: src/.env.sample [hunk 14]
```diff
@@ -147,5 +147,14 @@ RESOURCE_TYPES="projects,observations,observation_with_rubrics,surveys,programs"
...
+
+#kafka topic for publish
+PROJECT_PUBLISH_KAFKA_TOPIC=publishtopic
```
Comment (aks30) [inline src/.env.sample:157]: @priyanka-TL Let's introduce Kafka message on/off like we have here and based on that topic is optional.https://github.com/project-sunbird/ml-core-service/blob/release-8.0.0/.env.sample

File: src/services/projects.js [hunk 15]
```diff
@@ -608,7 +601,70 @@ module.exports = class ProjectsHelper {
...
+
+			if (!orgConfigList[common.PROJECT]) {
+				updateObj.status = common.PUBLISHED
+				updateObj.published_on = new Date()
+
+				//update resource with published_on and status
+				await resourceQueries.updateOne({ id: projectData.id }, updateObj)
+
+				//publish the resource in consumption side
+				if (process.env.CONSUMPTION_SERVICE != common.SELF) {
```
Comment (aks30) [inline src/services/projects.js]: @priyanka-TL Create a function which will take a resource id and push it to Kafka and  API. It will be used here and in some other place as well.

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 30 July

File: src/.env.sample (hunk 13 above)
Comment (priyanka-TL) [reply src/.env.sample]: @aks30 if we publish to both method then duplicate template will create. So can we go publish through kafka or api ?

File: src/services/comments.js [hunk 16]
```diff
@@ -32,6 +27,13 @@ module.exports = class ProjectsHelper {
 				})
 			}
 
+			//update the comment
+			if (bodyData.status === common.STATUS_RESOLVED || !bodyData.resolved_by) {
+				bodyData.resolved_by = loggedInUserId
+				bodyData.resolved_at = new Date()
+				bodyData.status = common.STATUS_RESOLVED
```
Comment (adithyadinesh0412) [inline src/services/comments.js]: It is already checking the status right? do we have to assign it again? @priyanka-TL

Comment (priyanka-TL) [reply src/services/comments.js]: its optional they can pass either status or resolved_by

File: src/services/comments.js [hunk 17]
```diff
@@ -32,6 +27,13 @@ module.exports = class ProjectsHelper {
...
+			}
+
 			const filter = {
 				id: comment_id,
 				resource_id: resource_id,
```
Comment (adithyadinesh0412) [inline src/services/comments.js]: Pass resource_id first as it is the DC ibn comments table. @priyanka-TL

File: src/database/migrations/20240513094933-create-comments-table.js
Comment (priyanka-TL) [reply src/database/migrations/20240513094933-create-comments-table.js:1]: yes

File: src/services/comments.js (hunk 17 above)
Comment (priyanka-TL) [reply src/services/comments.js]: id is the primary key here resource_id is dc but not a part of primary key

File: src/.env.backup [hunk 18]
```diff
@@ -21,9 +21,6 @@ KAFKA_URL = localhost:9092
 # Kafka group to which consumer belongs
 KAFKA_GROUP_ID ="selfcreation"
 
-# Kafka topic to push notification data
```
Comment (aks30) [inline src/.env.backup:24]: @priyanka-TL do we need this backup file?

File: src/constants/common.js [hunk 19]
```diff
@@ -19,7 +19,10 @@ module.exports = {
 		JWT_ONLY: 'jwt_only',
 		USER_SERVICE: 'user_service_authenticated',
 	},
-	internalAccessUrls: [process.env.APPLICATION_BASE_URL + 'v1/cloud-services/file/fetchJsonFromCloud'],
+	internalAccessUrls: [
+		process.env.APPLICATION_BASE_URL + 'v1/cloud-services/file/fetchJsonFromCloud',
```
Comment (aks30) [inline src/constants/common.js]: @priyanka-TL why do we need this API in internal access? Who will use it and when?

File: src/services/comments.js (hunk 17 above)
Comment (adithyadinesh0412) [reply src/services/comments.js]: you should pass the resource_id (distribution column) first to quickly locate the shard, followed by the id (primary key) to quickly locate the specific row within that shard.
This will be more optimised and easy to access if we pass the DC first, as comments table can have huge volume of data we can go for this I guess. 
@priyanka-TL

File: src/services/projects.js [hunk 20]
```diff
@@ -665,15 +656,48 @@ module.exports = class ProjectsHelper {
 				await reviewsQueries.bulkCreate(reviewsData)
 			}
 
+			//update the reviews and resource status
+			let resourceStatus = common.RESOURCE_STATUS_SUBMITTED
```
Comment (aks30) [inline src/services/projects.js:664]: @priyanka-TL In which scenario do we have to run this code? Updating reviews when calling details?

File: src/services/comments.js [hunk 21]
```diff
@@ -32,6 +33,13 @@ module.exports = class ProjectsHelper {
 				})
 			}
 
+			//update the comment
+			if (bodyData.status === common.STATUS_RESOLVED || !bodyData.resolved_by) {
+				bodyData.resolved_by = loggedInUserId
```
Comment (aks30) [inline src/services/comments.js]: @priyanka-TL Resolved by will always come from token right?

File: src/services/comments.js [hunk 22]
```diff
@@ -70,23 +78,23 @@ module.exports = class ProjectsHelper {
...
 	/**
-	 *  comment list
+	 * Comment list
 	 * @method
 	 * @name list
 	 * @param {Object} req - request data.
 	 * @returns {JSON} - comment list
 	 */
-	static async list(resource_id, organization_id) {
+	static async list(resource_id, page_value = '', context = '', loggedInUserId, organization_id) {
```
Comment (aks30) [inline src/services/comments.js]: @priyanka-TL We should just call it userId in the argument. Also update the function signature with correct description and covering all arguments.

File: src/services/resource.js [hunk 23]
```diff
@@ -626,6 +630,119 @@ module.exports = class resourceHelper {
...
+							}
+						}
+
+						result = { ...result, ...resultData }
+					}
+				}
+			}
+
+			//get organization details
+			let organizationDetails = await userRequests.fetchDefaultOrgDetails(resource.organization_id)
```
Comment (aks30) [inline src/services/resource.js]: @priyanka-TL why do we have function called "fetchDefaultOrgDetails" it should be just "fetchOrg". Also it has to be via org extension first.

File: src/services/resource.js [hunk 24]
```diff
@@ -626,6 +630,119 @@ module.exports = class resourceHelper {
...
+					response.result &&
+					Object.keys(response.result).length > 0
+				) {
+					//modify the response as label value pair
+					let resultData = response.result
+
+					//get all entity types with entities
+					let entityTypes = await entityModelMappingQuery.findEntityTypesAndEntities(
+						{
+							model: resource.type,
```
Comment (aks30) [inline src/services/resource.js]: @priyanka-TL Don't  we need for other models such as task and subtask?

File: src/services/resource.js [hunk 25]
```diff
@@ -626,6 +630,119 @@ module.exports = class resourceHelper {
...
+						const entityTypeMap = entityTypes.reduce((map, type) => {
+							if (type.has_entities && Array.isArray(type.entities) && type.entities.length > 0) {
+								map[type.value] = type.entities
+									.filter((entity) => entity.status === common.STATUS_ACTIVE)
+									.map((entity) => ({ label: entity.label, value: entity.value.toLowerCase() }))
+							}
+							return map
+						}, {})
+
+						for (let entityType of entityTypes) {
```
Comment (aks30) [inline src/services/resource.js]: @priyanka-TL Can you help understand what this leads to? Can we do it in parallel.

Also, we need a different API to list and search with pagination to get all entities for a given entity type. Rakesh has developed the same.

File: src/services/resource.js [hunk 26]
```diff
@@ -877,4 +1029,80 @@ module.exports = class resourceHelper {
...
 	}
+
+	/**
+	 * Check for direct publish without review
+	 * @method
+	 * @name isReviewMandatory
+	 * @returns {Boolean} - Review required or not
+	 */
+	static async isReviewMandatory(resourceType, organizationId) {
+		const orgConfig = await configService.list(organizationId)
```
Comment (aks30) [inline src/services/resource.js]: @priyanka-TL this has to be from org and not config.

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 6 Aug.

File: src/constants/common.js (hunk 19 above)
Comment (priyanka-TL) [reply src/constants/common.js]: fetchJsonFromCloud we added it to validate the json which is store in cloud, Previously detail api was only available to the creator of the project. Now its available to everyone. So i will remove fetchJsonFromCloud api

File: src/services/projects.js (hunk 20 above)
Comment (priyanka-TL) [reply src/services/projects.js:664]: @aks30  If the creator is submitting the resource for the second time, the resource status should be set to "IN_REVIEW" instead of "SUBMITTED." Additionally, we need to update the status in the reviews table to "CHANGES_UPDATED" for any reviews that were previously marked as "REQUESTED_FOR_CHANGES."

File: src/services/resource.js (hunk 24 above)
Comment (priyanka-TL) [reply src/services/resource.js]: Not required as of now. task and subtask dont have entities

File: src/services/resource.js (hunk 25 above)
Comment (priyanka-TL) [reply src/services/resource.js]: This will convert as label value pair. here we are listing entities based on model. I have updated this to parallel

**Review by adithyadinesh0412 — APPROVED** [review]
Can be approved from my side after changes suggested by @aks30  is completed. No more further comments from my side. Reviewed all changes I suggested previously.

File: src/.env.sample (hunk 14 above)
Comment (aks30) [reply src/.env.sample:157]: @priyanka-TL To include feature level - ON/OFF setting.

File: src/database/queries/reviews.js [hunk 27]
```diff
@@ -36,10 +36,59 @@ exports.distinctResources = async (filter, attributes) => {
...
+			{ returning: true }
+		)
 
 		return res
 	} catch (error) {
 		return error
 	}
 }
+
+exports.reviewsCount = async (filter) => {
```
Comment (aks30) [inline src/database/queries/reviews.js]: @priyanka-TL Rename it to count

File: src/helpers/getDefaultOrgId.js [hunk 28]
```diff
@@ -3,7 +3,7 @@ const userRequests = require('@requests/user')
 const utils = require('@generics/utils')
 exports.getDefaultOrgId = async () => {
 	try {
-		let defaultOrgDetails = await userRequests.fetchDefaultOrgDetails(process.env.DEFAULT_ORGANISATION_CODE)
+		let defaultOrgDetails = await userRequests.fetchOrg(process.env.DEFAULT_ORGANISATION_CODE)
```
Comment (aks30) [inline src/helpers/getDefaultOrgId.js]: @priyanka-TL As discussed we can get default org id from env rather than reaching out to user service.

File: src/services/comments.js [hunk 29]
```diff
@@ -1,29 +1,33 @@
...
+ */
 const httpStatusCode = require('@generics/http-status')
-const commentQueries = require('@database/queries/comment')
+const commentQueries = require('@database/queries/comments')
 const responses = require('@helpers/responses')
 const common = require('@constants/common')
 const userRequests = require('@requests/user')
 const _ = require('lodash')
 
 module.exports = class ProjectsHelper {
```
Comment (aks30) [inline src/services/comments.js]: @priyanka-TL Class name is wrong.

File: src/services/resource.js (hunk 24 above)
Comment (aks30) [reply src/services/resource.js]: @priyanka-TL As discussed please add for future use case.

File: src/services/reviews.js [hunk 30]
```diff
@@ -0,0 +1,447 @@
...
+	 * Update review.
+	 * @method
+	 * @name update
+	 * @param {Object} bodyData - review body data.
+	 * @param {Integer} resourceId - resource id.
+	 * @param {String} loggedInUserId - logged in user id.
+	 * @returns {JSON} - review updated response.
+	 */
+
+	static async update(resourceId, bodyData, loggedInUserId, orgId) {
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL To update function signature.

File: src/services/reviews.js [hunk 31]
```diff
@@ -0,0 +1,447 @@
...
+]
+
+async function handleComments(comments, resourceId, loggedInUserId) {
+	try {
+		// Normalize comments to an array if it's a single object
+		if (!Array.isArray(comments)) {
+			comments = [comments]
+		}
+		// handle adding or updating comments
+		for (const comment of comments) {
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL - We should not make queries in loop.

File: src/services/reviews.js [hunk 32]
```diff
@@ -0,0 +1,447 @@
...
+const _notAllowedReviewStatus = [
+	common.REVIEW_STATUS_STARTED,
+	common.RESOURCE_STATUS_REJECTED,
+	common.REVIEW_STATUS_REQUESTED_FOR_CHANGES,
+	common.REVIEW_STATUS_CHANGES_UPDATED,
+	common.REVIEW_STATUS_INPROGRESS,
+	common.REVIEW_STATUS_REJECTED_AND_REPORTED,
+]
+
+async function handleComments(comments, resourceId, loggedInUserId) {
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL - Function signature is missing

File: src/services/reviews.js [hunk 33]
```diff
@@ -0,0 +1,447 @@
...
+	 * @name handleUpdateReview
+	 * @param {Integer} reviewId - review id.
+	 * @param {Object} bodyData - review body data.
+	 * @param {Integer} resourceId - resource id.
+	 * @param {String} loggedInUserId - logged in user id.
+	 * @param {Array} comments - comments.
+	 * @param {String} organization_id - org id.
+	 * @returns {JSON} - review updated response.
+	 */
+	static async handleUpdateReview(reviewId, bodyData, resourceId, loggedInUserId, comments = [], organization_id) {
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL As discussed we will go with 3 different APIs - approve/update/reject&report

File: src/services/reviews.js [hunk 34]
```diff
@@ -0,0 +1,447 @@
...
+		try {
+			//update reviews table
+			await reviewsQueries.update({ id: reviewId, organization_id: organization_id }, { status: bodyData.status })
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL One can update a resource only if its in review.

File: src/services/reviews.js [hunk 35]
```diff
@@ -0,0 +1,447 @@
...
+		loggedInUserId,
+		comments = [],
+		minApproval,
+		resourceOrgId,
+		reviewOrgId
+	) {
+		try {
+			// handle approval of a review
+			let publishResource = false
+			await reviewsQueries.update(
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL One can approve a resource only if its in review.

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 7 Aug

File: src/database/queries/comments.js [hunk 36]
```diff
@@ -0,0 +1,111 @@
...
+				where: filter,
+				raw: true,
+			})
+			return comments
+		} catch (error) {
+			throw error
+		}
+	}
+
+	static async commentList(resourceId, userId, page_value, context) {
```
Comment (aks30) [inline src/database/queries/comments.js]: @priyanka-TL Just call it list.

File: src/services/reviews.js [hunk 37]
```diff
@@ -0,0 +1,534 @@
...
+			}
+
+			return publishResource
+		} catch (error) {
+			throw error
+		}
+	}
+}
+
+const _notAllowedStatusForReview = [
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL Please add comments on what this list is and where its used. The name is not very clear, consider changing it.

File: src/services/reviews.js [hunk 38]
```diff
@@ -0,0 +1,534 @@
...
+	common.RESOURCE_STATUS_REJECTED,
+	common.RESOURCE_STATUS_REJECTED_AND_REPORTED,
+	common.RESOURCE_STATUS_PUBLISHED,
+]
+
+const _notAllowedReviewStatus = [
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL Please add comments on what this list is and where its used. The name is not very clear, consider changing it.

File: src/services/reviews.js [hunk 39]
```diff
@@ -0,0 +1,534 @@
...
+	 * @name validateReview
+	 * @param {Integer} resourceId - resource id.
+	 * @param {String} userId - logged in user id.
+	 * @returns {JSON} - review updated response.
+	 */
+
+	static async validateReview(resourceId, userId) {
+		try {
+			//get resource details
+			let resourceDetails = await resourceService.getDetails(resourceId)
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL All these various blocks should be different functions, each will take only the necessary resource information. The calling function will call these functions with relevant information as needed and handle the responses accordingly.

**Review by aks30 — CHANGES_REQUESTED** [review]
Reviewed on 8 Aug

File: src/services/reviews.js [hunk 40]
```diff
@@ -0,0 +1,672 @@
...
+				})
+			}
+
+			await reviewResourceQueries.create(_.omit(reviewData, [common.STATUS]))
+
+			// Update resource table data
+			let updateData = {
+				status: common.RESOURCE_STATUS_IN_REVIEW,
+				last_reviewed_on: new Date(),
+				...(updateNextLevel && { next_stage: nextStage + 1 }),
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL We should update the next level only in approve flow.

File: src/services/reviews.js [hunk 41]
```diff
@@ -0,0 +1,672 @@
...
+
+	/**
+	 * Validating the Resource status
+	 * @method
+	 * @name validateResourceStatus
+	 * @param {String} status - resource status
+	 * @returns {JSON} - return error if any
+	 */
+
+	static async validateResourceStatus(status) {
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL change function name to isResourceAvailableForReview, should return true/false.

File: src/services/reviews.js [hunk 42]
```diff
@@ -0,0 +1,672 @@
...
+	 * Validate the reviewer level matches the resource level
+	 * @method
+	 * @name validateNextLevel
+	 * @param {Array<Object>} roles - user roles
+	 * @param {String} userOrgId - organization Id
+	 * @param {String} resourceType - Resource type
+	 * @param {Integer} currentLevel - current stage of resource
+	 * @returns {JSON} - return error if any
+	 */
+	static async validateNextLevel(roles, userOrgId, resourceType, currentLevel) {
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL Change function name to canUserReviewAResource, should return true/false.

File: src/services/reviews.js [hunk 43]
```diff
@@ -0,0 +1,672 @@
...
+		//get the roles
+		const userRoleTitles = utils.getUniqueElements(roles.map((item) => item.title))
+		//validate the level
+		const resourceWiseLevels = await resourceService.fetchReviewLevels(userOrgId, userRoleTitles, [resourceType])
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL Add sample arguments, better descripton and sample response of this function. Try to come up with a better name.

File: src/services/reviews.js [hunk 44]
```diff
@@ -0,0 +1,672 @@
...
+			if (!resource?.id) {
+				return responses.failureResponse({
+					message: 'RESOURCE_NOT_FOUND',
+					statusCode: httpStatusCode.bad_request,
+					responseCode: 'CLIENT_ERROR',
+				})
+			}
+
+			//validate resource
+			let validateReview = await this.validateReview(resourceId, userId, resource.status, orgId)
```
Comment (aks30) [inline src/services/reviews.js]: @priyanka-TL Change this variable to ongoingReviw. Add description and sample req/resp

**Review by aks30 — CHANGES_REQUESTED** [review]
Add more descriptive comments.
